PWGJE/TableProducer/jettaggerhf.cxx

```diff
@@ -21,64 +22,113 @@
+        vecSignedIP3Ds.push_back(geoSign * TMath::Abs(track.dcaXYZ()) / TMath::Sqrt(track.sigmaDcaXYZ2()) * JetTaggingUtilities::cmTomum);
+      }
+      constTableData(taggingTableData.lastIndex(), trackconst);
+      constTCTableData(taggingTableData.lastIndex(), 0, jet.pt(), jet.eta(), jet.phi(), vecGeoSign, vecTrackPt, vecTrackEta, vecTrackPhi, vecSignedIP2D, vecSignedIP2Ds, vecSignedIP3D, vecSignedIP3Ds);
     }
   }
   PROCESS_SWITCH(JetTaggerHFTask, processData, "Fill tagging decision for data jets", false);
 
-  void processMCD(aod::JCollision const& collision, JetTableMCD const& mcdjets, soa::Join<aod::JTracks, aod::McTrackLabels> const& tracks, aod::JMcParticles const& particles)
+  void processMCD(soa::Join<aod::JCollisions, aod::JCollisionPIs>::iterator const& jcollision, aod::Collisions&, JetTableMCD const& mcdjets, aod::JTracks&, JetTagTracksMC const& tracks, soa::Join<aod::JMcParticles, aod::JMcParticlePIs> const& particles)
   {
+    auto collision = jcollision.template collision_as<aod::Collisions>();
     for (auto& mcdjet : mcdjets) {
-
-      int origin = JetTaggingUtilities::mcdJetFromHFShower(mcdjet, tracks, particles, maxDeltaR);
+      typename JetTagTracksMC::iterator hftrack;
+      int origin = 0;
+      if (!doWShower)
+        origin = JetTaggingUtilities::mcdJetFromHFShower(mcdjet, tracks, particles, maxDeltaR);
+      else
+        origin = JetTaggingUtilities::jetTrackFromHFShower(mcdjet, tracks, particles, hftrack);
       int algorithm1 = 0;
       int algorithm2 = 0;
       int algorithm3 = 0;
       // if (doAlgorithm1) algorithm1 = JetTaggingUtilities::Algorithm1((mcdjet, tracks);
       // if (doAlgorithm2) algorithm2 = JetTaggingUtilities::Algorithm2((mcdjet, tracks);
       // if (doAlgorithm3) algorithm3 = JetTaggingUtilities::Algorithm3((mcdjet, tracks);
       taggingTableMCD(origin, algorithm1, algorithm2, algorithm3);
+      std::vector<int> trackconst;
+      std::vector<int> vecGeoSign;
+      std::vector<float> vecTrackPt, vecTrackEta, vecTrackPhi, vecSignedIP2D, vecSignedIP2Ds, vecSignedIP3D, vecSignedIP3Ds;
+      for (auto& track : mcdjet.template tracks_as<JetTagTracksMC>()) {
+        int geoSign = JetTaggingUtilities::getGeoSign(collision, mcdjet, track);
+        trackconst.push_back(track.globalIndex());
+        vecGeoSign.push_back(geoSign);
+        vecTrackPt.push_back(track.pt());
+        vecTrackEta.push_back(track.eta());
+        vecTrackPhi.push_back(track.phi());
+        vecSignedIP2D.push_back(track.dcaXY() / TMath::Sqrt(track.sigmaDcaXY2()));
+        vecSignedIP2Ds.push_back(geoSign * TMath::Abs(track.dcaXY()) / TMath::Sqrt(track.sigmaDcaXY2()));
+        vecSignedIP3D.push_back(track.dcaXYZ() / TMath::Sqrt(track.sigmaDcaXYZ2()));
+        vecSignedIP3Ds.push_back(geoSign * TMath::Abs(track.dcaXYZ()) / TMath::Sqrt(track.sigmaDcaXYZ2()));
+      }
+      constTableMCD(taggingTableMCD.lastIndex(), trackconst);
```

Comment:
since these tables can be joined to the taggingTableMCD I dont think you need to the indices. you can just  join them later